### PR TITLE
Add snapshot/restore support for virtio-mmio

### PR DESCRIPTION
**Pull request by sboeuf**
This pull request introduces snapshot/restore support to all virtio devices using the virtio-mmio transport layer. It only contains support for virtio-rng and virtio-blk for now, but more devices will be added through follow up pull requests.

### CODE REVIEW

**Comment by yisun-git** [comment]
Overall, LGTM. Thanks!

File: vm-virtio/src/transport/mmio.rs [hunk 1]
```diff
@@ -58,6 +62,14 @@ impl VirtioInterrupt for VirtioInterruptIntx {
...
 }
 
+#[derive(Serialize, Deserialize)]
+pub struct VirtioMmioDeviceState {
+    pub device_activated: bool,
+    pub driver_status: u32,
+    pub interrupt_status: usize,
+    pub queues: Vec<Queue>,
+}
+
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply vm-virtio/src/transport/mmio.rs:72]: Oh good catch! I added a few fields I was missing :)

File: vm-virtio/src/transport/mmio.rs [hunk 2]
```diff
@@ -123,6 +135,34 @@ impl MmioDevice {
...
+        // Update virtqueues indexes for both available and used rings.
+        if let Some(mem) = self.mem.as_ref() {
+            let mem = mem.memory();
+            for queue in self.queues.iter_mut() {
+                queue.update_avail_index_from_memory(&mem);
+                queue.update_used_index_from_memory(&mem);
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply vm-virtio/src/transport/mmio.rs:162]: :)

File: vm-virtio/src/transport/mmio.rs [hunk 3]
```diff
@@ -354,6 +407,84 @@ impl Pausable for MmioDevice {
...
+                        self.device
+                            .lock()
+                            .unwrap()
+                            .activate(
+                                mem,
+                                interrupt_cb,
+                                self.queues.clone(),
+                                self.queue_evts.split_off(0),
+                            )
+                            .expect("Failed to activate device");
```
Comment (rbradford) [inline vm-virtio/src/transport/mmio.rs:475]: nit: Can we replace this panic with an error?

File: vmm/src/device_manager.rs [hunk 4]
```diff
@@ -1306,8 +1310,8 @@ impl DeviceManager {
                 ))
             };
             self.add_migratable_device(
-                Arc::clone(&virtio_net_device) as Arc<Mutex<dyn Migratable>>
-            );
+                            Arc::clone(&virtio_net_device) as Arc<Mutex<dyn Migratable>>
+                        );
```
Comment (rbradford) [inline vmm/src/device_manager.rs:1314]: Weird whitespace change! (has cargo fmt changed the way it works...?)

File: vm-virtio/src/transport/mmio.rs (hunk 3 above)
Comment (sboeuf) [reply vm-virtio/src/transport/mmio.rs:475]: You mean `map_err()`, right?

File: vmm/src/device_manager.rs (hunk 4 above)
Comment (sboeuf) [reply vmm/src/device_manager.rs:1314]: Oh yeah that's weird, I didn't spot that, let me check.

Comment (sboeuf) [reply vmm/src/device_manager.rs]: I changed it :)

File: vm-virtio/src/transport/mmio.rs (hunk 3 above)
Comment (sboeuf) [reply vm-virtio/src/transport/mmio.rs]: Done

**Pull request merged**